maskit/masks.py

```diff
@@ -0,0 +1,296 @@
+        raise NotImplementedError
+
+    @property
+    def mask(self):
+        """
+        Returns the encapsulated :py:attr:`~.mask`
+        """
+        return self._mask
+
+    def apply_mask(self, values: np.ndarray):
+        """
+        Applies the encapsulated py:attr:`~.mask` to the given ``values``.
+        Note that the values should have the same shape as the py:attr:`~.mask`.
+
+        :param values: Values where the mask should be applied to
+        """
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """Resets the mask to not mask anything."""
+        self._mask = np.zeros_like(self._mask, dtype=bool, requires_grad=False)
+
+    def perturb(
+        self,
+        amount: Optional[int] = None,
+        mode: PerturbationMode = PerturbationMode.INVERT,
+    ):
+        """
+        Perturbs the MaskedObject by the given ``mode`` of type
+        :py:class:`~.PerturbationMode` ``amount`` times. If no amount is given,
+        that is ``amount=None``, a random ``amount`` is determined given by the
+        actual size of the py:attr:`~.mask`. The ``amount`` is automatically
+        limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to perturb, defaults to None
+        :param mode: How to perturb, defaults to PerturbationMode.INVERT
+        :raises AssertionError: Raised for negative amounts
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert (
+            amount is None or amount >= 0
+        ), "Negative values are not supported, please use PerturbationMode.REMOVE"
+        if amount == 0:
+            return
+        count = (
+            abs(amount) if amount is not None else rand.randrange(0, self._mask.size)
+        )
+        if mode == PerturbationMode.ADD:
+            indices = np.argwhere(~self._mask)
+        elif mode == PerturbationMode.INVERT:
+            indices = np.array([list(index) for index in np.ndindex(*self._mask.shape)])
+        elif mode == PerturbationMode.REMOVE:
+            indices = np.argwhere(self._mask)
+        else:
+            raise NotImplementedError(f"The perturbation mode {mode} is not supported")
+        if len(indices) == 0:
+            return
+        indices = tuple(
+            zip(
+                *indices[
+                    np.random.choice(
+                        len(indices), min(count, len(indices)), replace=False
+                    )
+                ]
+            )
+        )
+        self._mask[indices] = ~self._mask[indices]
+
+    def copy(self) -> "MaskedObject":
+        """Returns a copy of the current MaskedObject."""
+        clone = object.__new__(type(self))
+        clone._mask = self._mask.copy()
+        return clone
+
+
+class MaskedParameter(MaskedObject):
+    """
+    A MaskedParameter encapsulates not only the :py:attr:`~.mask` but also the
+    according :py:attr:`~.parameters` being masked.
+    """
+
```

Comment:
Since ``MaskedObject`` uses ``__slots__``, strongly consider for subclasses to use ``__slots__`` as well. You only need the additional slots, e.g.
```suggestion
    __slots__ = ("_parameters",)
```